pkg/apis/csi.min.io/v1alpha1/types.go

```diff
@@ -0,0 +1,123 @@
+
+// StorageTopologyList is a list of StorageTopology objects.
+type StorageTopologyList struct {
+	metav1.TypeMeta `json:",inline"`
+	// metdata is the standard list metadata.
+	// +optional
+	metav1.ListMeta `json:"metadata"`
+
+	// items is the list of Storage Topology objects.
+	Items []StorageTopology `json:"items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// StorageTopology defines the layout of storage infrastructure.
+type StorageTopology struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the storage topology
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status.
+	// +optional
+	Spec StorageTopologySpec `json:"spec,omitempty"`
+	// StorageTopologyStatus defines the observed state of StorageTopology
+	// +optional
+	Status StorageTopologyStatus `json:"status,omitempty"`
+}
+
+// StorageTopologyLayout describes the nodes and drives that will be used across the cluster.
+type StorageTopologyLayout struct {
+	// Labels of the nodes that will participate
+	NodeLabels map[string]string `json:"nodeLabels,omitempty"`
+	// Mount Path for the volumes that will be used
+	Path string `json:"string"`
+}
+
+// StorageTopologySpec is the spec of the StorageTopology.
+type StorageTopologySpec struct {
+	// Layout of the drives across hosts
+	Layout StorageTopologyLayout `json:"layout"`
+	// Layout of the drives across hosts
+	// +optional
+	FsType string `json:"fstype,omitempty"`
+	// Mount Options
+	// +optional
+	MountOptions []string `json:"mount_options,omitempty"`
+	// Resource Limits
+	// +optional
+	ResourceLimits StorageTopologyResourceLimit `json:"resource_limits,omitempty"`
+	// Reclaim Policy
+	// +optional
+	ReclaimPolicy string `json:"reclaim_policy,omitempty"`
+}
+
+type StorageTopologyResourceLimit struct {
+	// Storage
+	// +optional
+	Storage resource.Quantity `json:"storage,omitempty"`
+	// Volumes
+	// +optional
+	Volumes int32 `json:"volumes,omitempty"`
+}
+
+// StorageTopologyStatus is the state of the StorageTopology
+type StorageTopologyStatus struct {
+	// Nodes the storage topology covers
+	// +optional
+	Nodes []string `json:"accessibleNodes,omitempty"`
+	// Total Capacity of the storage topology across all nodes
+	// +optional
+	Capacity StorageTopologyStatusCapacity `json:"capacity,omitempty"`
+	// Present only when there is an error.
+	// +optional
+	Error StorageTopologyStatusError `json:"error,omitempty"`
+}
+
+type StorageTopologyStatusCapacity struct {
+	// Free Space of the storage topology
+	// +optional
+	FreeSpace resource.Quantity `json:"freeSpace,omitempty"`
+	// Total capacity of the storage topology
+	// +optional
+	Total resource.Quantity `json:"total,omitempty"`
+}
```

Comment:
needs a field for the number of allocated volumes

Reply:
call it `volumeAllocations`

Reply:
just an integer?

Reply:
It should go into `StorageTopologyStatus ` field

Reply:
yeah, just an integer

Reply:
done